### DISCUSSION TITLE
Thank you & small Python app demo

### DISCUSSION THREAD

**Discussion by bluenote10** [khonsulabs/gooey#157] [Show and tell]
I just wanted to take the opportunity to say thank you! I've been following your work/blog a little bit, which is always very interesting and inspiring. I also share your excitement regarding cushy :wink:

In case you are interested: I've implemented some [prototypical Python bindings](https://github.com/bluenote10/RustExperiments/tree/master/simple_app), allowing to write tiny (fully type-safe!) Python apps like this:

```py
import numpy as np

import simple_app
from simple_app import Plot, Slider

elements = [
    (slider_a := Slider("a", -10.0, 0.5, 10.0)),
    (slider_b := Slider("b", -10.0, 0.5, 10.0)),
    (slider_c := Slider("c", -10.0, 0.5, 10.0)),
]


def callback() -> list[Plot]:
    print(f"{slider_a.value=} {slider_b.value=} {slider_c.value=}")
    a = slider_a.value
    b = slider_b.value
    c = slider_c.value
    xs = np.linspace(-10.0, 10.0, 100)
    ys = a * xs**2 + b * xs + c
    return [Plot(xs, ys, x_limits=(-10, +10), y_limits=(-10, +10))]


simple_app.run(elements, callback)
```

Which produces something like this:

https://github.com/khonsulabs/cushy/assets/3620703/ae3ddde3-4ded-4886-b80a-bf857acdb989

The bindings aren't meant to be general purpose, but rather offering a domain specific interface for typical algorithmic experimentation I'm often faced with. In a sense, Desmos-like apps, but powered by Python allowing to visualize algorithmic ideas, or just arbitrary dynamic data exploration etc. Of course the feature set is far from complete / usable.

What's pretty cool: The rendering performance of PyO3 + cushy with reactive callbacks into Python is very promising. In the Python ecosystem there are a number of browser-based app frameworks that allow to write similar apps, but the frontend/backend communication always feels so slow. In comparison, the small cushy demo feels very snappy; a refreshing alternative to all the browser stuff!

**Comment by ecton** [comment]
Wow, this is really awesome! I'm really happy you were able to get something like this working and that you've been pleased with the performance too. Thank you for sharing!
